qiskit/pulse/library/parametric_pulses.py

```diff
@@ -70,6 +71,14 @@ def __init__(
                              amplitude is constrained to 1.
         """
         super().__init__(duration=duration, name=name, limit_amplitude=limit_amplitude)
+
+        warnings.warn(
+            "ParametricPulse and its subclass have been deprecated and will be replaced with "
+            "SymbolicPulse and its subclass because of QPY serialization support. "
+            "See qiskit.pulse.library.symbolic_pulses for details.",
+            PendingDeprecationWarning,
```

Comment:
Data format in the communication layer is one of the important aspects of the symbolic pulse, however, for us, the most important feature is the serialization of calibration database, where users may actively introduce new pulse shapes with new calibration techniques. 

I tend to disagree with supporting both `ParametricPulse` and `SymbolicPulse` from viewpoint of maintenance overhead. This means we need to keep double amount of unit tests. Since we already have `Schedule` and `ScheduleBlock`, this will make situation more complicated, i.e. there are 4 different combination of data models to create a pulse program. 

> I would have expected to be able to pass a pulse name and set of parameters and as long as the pulse name was in the backend's run config for it to be okay.

This is why we need symbolic pulse. Indeed how this code is used is unclear now, because job submission mechanism is totally offloaded to provider. Qiskit just passes python object to the provider. If one introduces new pulse shape in the backend config, he/she should update the provider code together with parametric pulse library in Qiskit. And user need to wait for next release to enjoy backend-supported parametric pulse. I think this is why DCX (target) pulse is still provided as waveform.

I agree keeping current mechanism of name + parameters for predefined pulses. This will help us to reduce data volume of QPY binary. Regarding the serialization, I plan to write symbolic equations in the program header, instead of duplicating it in every appearance of play instruction. For example, in the circuit we have
https://github.com/Qiskit/qiskit-terra/blob/04807a13a7ba53d97c37bb092324e419296e3fba/qiskit/qpy/binary_io/circuits.py#L591-L607

If we only save equations of custom pulses, we can compactly represent the program (i.e. DRAG is very often used but constraints is quite big object). Same mechanism exists in circuit IO.

Reply:
Yeah I was thinking to use `pulse_type` for this purpose but this induces a breaking API change. Because this camel case name is used everywhere, e.g. printing schedule would give
```
Schedule((0, Play(GaussianSquare(duration=800, amp=(0.1+0j), sigma=40, width=544), DriveChannel(0))), name="sched17")
```
If the symbolic pulse needs to use conventional names in `parametric_pulses`, then this will become
```
Schedule((0, Play(gaussian_square(duration=800, amp=(0.1+0j), sigma=40, width=544), DriveChannel(0))), name="sched17")
```
which may confuse the users. This also changes labels in the pulse visualizer and `repr` of pulse instances. Also we still need this Enum to convert backend's CmdDef back to the schedules (i.e. construction of instmap) because we need to extract envelope expressions from somewhere in the pulse module unless the backend replaces CmdDef with QPY schedules. I agree this approach is still effective to simplify the logic of schedule -> Qobj.